src/lib/homestore_backend/pg_blob_iterator.cpp

```diff
@@ -36,102 +82,178 @@ PG* HSHomeObject::PGBlobIterator::get_pg_metadata() {
-    //                        builder.CreateVector(b.blob.body.bytes(), b.blob.body.size())));
-    // }
-    // builder.FinishSizePrefixed(
-    //     CreateResyncBlobDataBatch(builder, builder.CreateVector(blob_entries), end_of_shard /* end_of_batch */));
-    // data_blob = sisl::io_blob_safe{builder.GetSize()};
-    // std::memcpy(data_blob.bytes(), builder.GetBufferPointer(), builder.GetSize());
+bool HSHomeObject::PGBlobIterator::create_blobs_snapshot_data(sisl::io_blob_safe& data_blob) {
+    std::vector< ::flatbuffers::Offset< ResyncBlobData > > blob_entries;
+    flatbuffers::FlatBufferBuilder builder;
+
+    bool end_of_shard = false;
+    uint64_t total_bytes = 0;
+    auto idx = (uint64_t)(last_end_blob_idx + 1);
+
+    while (total_bytes < max_batch_size && idx < cur_blob_list.size()) {
+        auto info = cur_blob_list[idx];
+        ResyncBlobState state = ResyncBlobState::NORMAL;
+        //handle deleted object
+        if (info.pbas == tombstone_pbas) {
+            state = ResyncBlobState::DELETED;
+            LOGD("Blob is deleted: shard_id={}, blob_id={}, blkid={}", info.shard_id, info.blob_id,
+                 info.pbas.to_string());
+            blob_entries.push_back(CreateResyncBlobDataDirect(builder, info.blob_id, (uint8_t)state, nullptr));
+            continue;
+        }
+        auto result = load_blob_data(info, state).get();
+        if (result.hasError() && result.error().code == BlobErrorCode::READ_FAILED) {
+            LOGW("Failed to retrieve blob for shard={} blob={} pbas={}, err={}", info.shard_id, info.blob_id,
+                 info.pbas.to_string(), result.error());
+            //TODO add metrics or events to track this
+            return false;
+        }
+
+        auto& blob = result.value();
+        std::vector< uint8_t > data(blob.cbytes(), blob.cbytes() + blob.size());
+        blob_entries.push_back(CreateResyncBlobDataDirect(builder, info.blob_id, (uint8_t)state, &data));
+        total_bytes += blob.size();
+    }
+    idx++;
+
+    if (idx == cur_blob_list.size()) { end_of_shard = true; }
+    cur_batch_blob_count = blob_entries.size();
+    builder.FinishSizePrefixed(
+        CreateResyncBlobDataBatchDirect(builder, &blob_entries, end_of_shard));
+    auto payload = sisl::io_blob_safe{builder.GetSize()};
+    std::memcpy(payload.bytes(), builder.GetBufferPointer(), builder.GetSize());
+    LOGD("create blobs snapshot data batch: shard_id={}, batch_num={}, total_bytes={}, blob_num={}, end_of_shard={}",
+         cur_shard_seq_num, cur_batch_num, total_bytes, blob_entries.size(), end_of_shard);
+
+    pack_resync_message(data_blob, SyncMessageType::SHARD_BATCH, payload);
+    return true;
 }
 
-bool HSHomeObject::PGBlobIterator::end_of_scan() const {
-    return max_shard_seq_num_ == 0 || cur_shard_seq_num_ > max_shard_seq_num_;
+void HSHomeObject::PGBlobIterator::pack_resync_message(sisl::io_blob_safe& dest_blob, SyncMessageType type,
```

Comment:
Please try to avoid the double copy from builder -> payload then payload -> dest_blob